docs: add OpenAPI specs

akash2237778 opened the pull request.

akash2237778 committed:
docs: Add OpenAPI specs

lvarin commented on broker.yaml:
```diff
@@ -0,0 +1,717 @@
+            type: string
+      responses:
+        '200':
+          description: 'List subcriptionId for all subscribed services'
```
This should be a list of subscription ids, right?

lvarin commented on broker.yaml:
```diff
@@ -0,0 +1,717 @@
+            type: string
+            description: URL on which CD is listening for image updates.
+            example: https://ec2-54-203-145-132.compute-1.amazonaws.com/update
+        accessToken:
```
In which occasions we want to show the `accessToken`? 

The broker creates the token and returns it to "operations" in the POST. It makes sense. But then, do we want to show it as well in the GET and PUT? It is fine for me, but I think it is something to consider.

akash2237778 replied:
Yes, exactly we need not show `accessToken` again. I'll update it. Thanks!

But it's not like that broker won't create token for "operations". 

@uniqueg : For the CD part, the tokens/secrets should be managed at the updater service and included when registering the deployment with the broker to ensure that calls from the broker to the updater include the CD token. 
https://elixir-cloud.slack.com/archives/C018MFNN6LS/p1622104620040100

lvarin reviewed (commented):
Only minor things

akash2237778 committed:
Minor fix

removed accessToken from GET /services/{subcriptionId}, Minor fix: show list at GET /services

akash2237778 committed:
Minor fix

Modified description in broker.yml and README.md

lvarin commented:
For me it looks good, what about you @uniqueg ?

uniqueg commented on README.md:
```diff
@@ -4,7 +4,7 @@
 ## Synopsis
 
-Broker is a decoupled, publish-subscribe-based continuous integration (CI) and continuous delivery (CD) microservice that allows developers to notify deploments of available updates, which can then autonomously decide what to do with them. Broker has a optional link between between the [ELIXIR Cloud Service Registry][elixir-cloud-registry] and the Broker.
+Broker is a decoupled, publish-subscribe-based continuous integration (CI) and continuous delivery (CD) microservice that allows developers to notify deploments of available updates, which can then autonomously decide what to do with them. Broker has an optional link with [ELIXIR Cloud Service Registry][elixir-cloud-registry].
```
- typo: `deployments` instead of `deploments`
- add `the`: `with the ELIXIR Cloud Service Registry`
- Broker itself is not decoupled, and neither does it do any CI/CD itself. Also, it is technically not the deployments themselves that decide what to do with updates, but the devops/sys admins. So perhaps better to write something like this:
 
```console
Broker leverages the publish-subscribe (pub/sub) pattern to effectively decouple
continuous integration (CI) from continuous deilivery (CD): developers can call
Broker in their CI pipelines to notify deployments of available updates
("publish"), while deployments can listen to update notifications ("subscribe")
and trigger specific CD pipelines set up by administrators.

Broker is particularly useful wherever deployments are managed by different
entities than those managing the underlying code bases, such as in federated
networks, in which services are deployed at multiple locations by different
organizations.
```

uniqueg replied:
Another general comment on writing markdown: try to not put more than 80 chars in a single line. This will make it easier to work with it (during editing, review etc). You can use editor plugins to help you with line wrapping.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+  title: 'ELIXIR Broker Registry API Specification'
+  version: 1.0.0
+  description: |
+    Broker is a decoupled, publish-subscribe-based continuous integration (CI) and continuous delivery (CD) microservice that allows developers to notify deploments of available updates, which can then autonomously decide what to do with them.
```
Update according to description above. Make sure to wrap over multiple lines to improve readability.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+    Broker is a decoupled, publish-subscribe-based continuous integration (CI) and continuous delivery (CD) microservice that allows developers to notify deploments of available updates, which can then autonomously decide what to do with them.
+  license:
+    name: 'Apache 2.0'
+    url: 'https://raw.githubusercontent.com/ga4gh-discovery/ga4gh-service-registry/develop/LICENSE'
```
Link to the version of the license in _this_ repository

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+    name: 'Apache 2.0'
+    url: 'https://raw.githubusercontent.com/ga4gh-discovery/ga4gh-service-registry/develop/LICENSE'
+  contact:
+    name: 'ELIXIR Cloud & AAI Members'
```
Remove `Members`

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+    url: 'https://raw.githubusercontent.com/ga4gh-discovery/ga4gh-service-registry/develop/LICENSE'
+  contact:
+    name: 'ELIXIR Cloud & AAI Members'
+    url: 'https://elixir-europe.github.io/cloud/categories/people.html'
```
This page is quite outdated. Better to link to <https://github.com/elixir-cloud-aai/elixir-cloud-aai>

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+    name: 'ELIXIR Cloud & AAI Members'
+    url: 'https://elixir-europe.github.io/cloud/categories/people.html'
+paths:
+  /projects:
```
If we cannot find a use case in which the broker could be used for anything other than Git repositories, I think I'd prefer `repositories`, as it is more specific/descriptive.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+                      type: string
+                      description: Used to verify later at /projects/{projects_id} if the request is from a trusted source.
+                      example: xxxxxxxxxxxxxx
+        '401':
```
Please include a `400/BadRequest` here and in all other endpoints.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+              schema:
+                type: object
+                properties:
+                    project_id:
```
I would keep property names short, as the grouping under `/projects` already makes them sufficiently specific: `id` and `access_token`

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+    url: 'https://elixir-europe.github.io/cloud/categories/people.html'
+paths:
+  /projects:
+    post:
```
Please add a `description` property for each endpoint to give a more detailed description of what an endpoint does. Use bullet point-style for the summary (as you already do) and phrase-like style for the description. An example description for this endpoint could be:

```
Register a project to enable publishing updates via the `/projects/{id}/builds` endpoint and allow administrators to subscribe to them.
```

More details can be added later as/if necessary.

See my comments below to understand why I renamed the update/builds endpoint.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+                      example: project_123
+                    project_access_token:
+                      type: string
+                      description: Used to verify later at /projects/{projects_id} if the request is from a trusted source.
```
Please be consistent with your descriptions/text. Either always use punctuation at the end, or never use it. I tend to think it's easier/clearer to always use punctuation and, accordingly, formulate text in a sentence-/phrase-like manner.

uniqueg replied:
Also, perhaps clearer to write:

```
Secret to be used to identify the source when publishing updates via `/projects/{id}/builds`.
```

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+                properties:
+                    project_id:
+                      type: string
+                      description: Unique id generated for each project
```
For consistency, it's perhaps better to use full phrases for each description. Also, best to avoid abbreviations (even seemingly obvious ones). Suggestion: `A unique identifier for the generated resource.` This description is also sufficiently generic that it could be reused in other places where we are perhaps creating resources, while still being specific enough if used together with a specific endpoint description.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+        default:
+          $ref: '#/components/responses/Error'
+    get:
+      summary: 'List all available projects.'
```
To keep this even more brief/bullet-point-style: `List available projects` as `available` already implies `all`.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+        - projects
+      responses:
+        '200':
+          description: 'List of all available projects'
```
Again, `available` implies `all`.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+          $ref: '#/components/responses/InternalServerError'
+        default:
+          $ref: '#/components/responses/Error'
+  /projects/{projectId}:
```
I would rename this `/projects/{id}/builds` as essentially you are providing here a different type of resource than those served at `/projects`. Instead, you can define a `GET` endpoint for `/projects/{id}` where you return details of a single project (as opposed to `GET /projects` which lists _all_ projects that a user has access to).

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+          schema:
+            type: string
+          description: Identifier generated for each project on POST
+        - in: cookie
```
I'm not sure whether the overhead of dealing with cookies is justified. For the beginning at least I would use a custom `header`

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+            type: string
+          description: Identifier generated for each project on POST
+        - in: cookie
+          name: Project Access Token
```
If using header, maybe use something like: ``X-Project-Access-Token`.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+          required: true
+          schema:
+            type: string
+          description: Identifier generated for each project on POST
```
This is not really clear. Better something like: "Identifier generated when first registering the project via `POST /projects`."

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+          $ref: '#/components/responses/Error'
+  /projects/{projectId}:
+    post:
+      summary: 'Registers a new build for the project'
```
For consistency with previous descriptions, better to always address the user, so `Register` instead of `Registers`. Also, for the `summary`, please use bullet point style, i.e., remove articles etc. For example, here: `Register new build for project`. Then in the `description` formulate as a phrase (and, if necessary, extend).

I won't comment these anymore from now on, but please apply all my comments on `summary`, `description` and consistency on all occurrences.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/BuildInfo'
```
For consistency with the `Project` schema, I'd just use `Build` here (or `BuildRegister`, see below).

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+                items:
+                  allOf:
+                    - $ref: '#/components/schemas/BuildInfo'
+                  properties:
```
For improved extensibility and clarity, it may be better to define two schemas `Build` and `BuildRegister`, then include these additional properties (and all of `BuildRegister`) in `Build` down in the schemas section.

The same may apply for `Project` / `ProjectRegister`.

FYI: OpenAPI has limited support for polymorphism etc., and in cases where the registration requires properties that are not to be included in responses, it may not be easy or possible at all to re-use schema parts in that way. So if we run into this, it may be necessary to duplicate parts of schemas.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+          $ref: '#/components/responses/InternalServerError'
+        default:
+          $ref: '#/components/responses/Error'
+    delete:
```
I guess this should go further up, not under `/projects/{id}/builds`.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+          $ref: '#/components/responses/InternalServerError'
+        default:
+          $ref: '#/components/responses/Error'  
+  /projects/{projectId}/{buildId}:
```
`/projects/{id}/builds/{buildId}`

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+        default:
+          $ref: '#/components/responses/Error'  
+  /projects/{projectId}/{buildId}:
+    put:
```
You could include a `PUT` method for `/projects/{id}` as well.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+          $ref: '#/components/responses/Error'  
+  /projects/{projectId}/{buildId}:
+    put:
+      summary: 'Update already created build'
```
How do you envision updating a build in practice? If you don't have a clear idea yet, you might want to leave this endpoint out, at least for now. In any case, `DELETE` and `POST` could always be used together for a similar effect.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+                    - $ref: '#/components/schemas/BuildInfo'
+                  properties:
+                      startedAt:
+                          type: number
```
Look here for the preferred specification of `date-time` in OpenAPI: https://swagger.io/docs/specification/data-models/data-types/#string

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+                          type: number
+                          example: 1622652950
+                          description: Finish time of latest build
+                      buildStatus:
```
Better to define a reusable enumerator: https://swagger.io/docs/specification/data-models/enums/

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+          $ref: '#/components/responses/InternalServerError'
+        default:
+          $ref: '#/components/responses/Error'
+    delete:
```
Conceptually, how do you envision the consequences of deleting a build? Would be best to open an issue to propose a behavior, then we can discuss this (as we haven't really done that before). Consider what happens/should happen if a build has not yet started, has already been started, some/all deployments have been notified etc. Please do the same for deleting/updating a project (and updating a build).

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+          $ref: '#/components/responses/InternalServerError'
+        default:
+          $ref: '#/components/responses/Error'  
+  /services:
```
I think this should really be called `/subscriptions` as this is what we really manage via this family of endpoints

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+      tags:
+        - services
+      parameters:
+        - in: cookie
```
See above for a discussion re: cookie vs header

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+                      type: string
+                      description: Unique id generated for each project
+                      example: project_123
+                    project_access_token:
```
I think we also need endpoints to manage tokens. You can create an issue for that and take care of it later.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+              allOf:
+                - $ref: '#/components/schemas/Service'
+              properties:
+                  accessToken:
```
Again, we'd need to manage these. Also, for consistency, use `snake_case`, not `camelCase` for properties.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+        default:
+          $ref: '#/components/responses/Error'
+    get:
+      summary: 'Shows build information'
```
I think this should rather just list metadata/info of a single subscription (the same stuff or more than what you get for calling `GET /subscriptions`.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+          $ref: '#/components/responses/InternalServerError'
+        default:
+          $ref: '#/components/responses/Error'  
+  /services/from-registry:
```
I'm not sure this should be a separate endpoint, but rather a property of the schema used to register a subscription. I would also leave this out for now - we can take care of the integration with the service registry later, because I think it will be clearer how exactly to do this once we have an MVP that works end-to-end (how to map service registry service properties to subscription properties? Is there information missing?). Best to open an issue for that.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+        default:
+          $ref: '#/components/responses/Error'  
+components:
+  securitySchemes:
```
Note that it may be difficult to post tokens to `POST /projects/{id}/builds` when used in a CI pipeline. Instead, we are planning to use the `access_token` secret to ensure that only trusted builds are triggered. Consequently, that endpoint (and possibly others) should probably not be protected by auth. Check the OpenAPI specs/docs to see how auth can be enabled for individual endpoints.

lvarin replied:
I am not sure about this. What is the usual way to make these kind of calls in a CI/CD environment? For `curl` or `python` sending header is not hard.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+      type: http
+      scheme: bearer
+      bearerFormat: JWT
+  responses:
```
Add `BadRequest`.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+            type: string
+            description: Unique id generated for each project
+            example: project_12
+        callbackUrl:
```
I think callbacks can be actually specified in OpenAPI3: https://swagger.io/docs/specification/callbacks/ In this way, I think, you can define the schema for the callback payload right in this specification!

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+  schemas:
+    Service:
+      type: object
+      properties:
```
I think this also needs a build type. Details still need to be discussed. Please open an issue and propose a solution on how a subscription can listen to just specific updates. You need to think this through from both the developer and admin point of view. The working proposal I think involved a set of defined types, I think `production`, `development`, `testing` to be used for stable new releases, development releases and any code change, respectively. However, preferably, we could also think of a more generic way of dealing with this, e.g., using branches or tags (which we could access via Git), given that this information is already specified in the `Build` schema (see below). Also consider other comments below. Basically, instead of defining just a fixed number of types, we could also use either "tags" (i.e., only subscribe to any new tags) or any branch name. And then those could be either manually defined as available types by the developer OR (probably better) automatically created for each repo.

Also, we would want to allow the administrators to specify an `audience` of people that they trust for code changes. If not specified, all verified commits would be automatically trusted. While we could leverage GitHub concepts here (organizations, teams, code owners, code managers etc), I think the only GitHub-independent solution would be a list of GPG public keys that developers use to sign their commits with.

Finally, I also think we need to include a commit SHA here (either via branch/commit or a tag, see below for the `HeadCommit` model proposal). The reasoning is the following: we can assume that the person registering the subscription is already running a service (or wants to run a service) that they know and trust. But this knowledge/trust can only be assumed for a certain commit, not any commits before and, more importantly, after. So when registering a subscription, the administrator would indicate that they trust the subscribed service at the indicated commit and would be interested to subscribe to updates, as long as those updates (1) were added _on top_ of the trusted commit, and (2) consist of only verified commits of the trusted audience. In a given build, then, the head commit should be used to keep track of the last trusted commit, so that for future code changes/updates, the broker could use that new head commit to verify only the new commits added on top (and so on).

uniqueg replied:
Just to add on: it may be even better to store all commits up to and including the indicated trusted head commit (and for all updates any future commits up to and including the new head commits), so that, e.g., renaming branches would not automatically invalidate a subscription, as the consistency of a branch would then depend on its constituing member commits and not the branch name.

lvarin replied:
What if history is rewritten? For example a password is committed to master/main.

lvarin replied:
Branches will share commits and feature branches will be rebased to master HEAD. I think makes storing all comits too complicated, and maybe it is better to check every time that every commit is GPG signed by an allowed person.

lvarin replied:
Another corner case is when part of the code is developed by a trusted developer that then leaves. Older commits should be accepted, but not newer ones. Then maybe we should store all commits.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+          type: string
+          description: Last commit sha till where repo needs to be checkout by git
+          example: '6b9899f'
+        imageTag:
```
I think we need to consider cases where a service uses more than one Docker image, so this should probably be a list or an object. Perhaps we should leave this for now and just open an issue to discuss (because we need to come up with a solution that allows the deployments to identify which pod/deployment should be updated with which image). This could easily become tricky, especially if we don't want to be locked in to k8s-based deployment environments.

uniqueg replied:
What we should probably account for though, even now, is that a `Dockerfile` may not always be called `Dockerfile`, and it may not always be located in the repository's root directory. So perhaps we should rename this property `image` and have a nested object inside with properties `tag` and `location` for now. Alterantively, to think ahead, you could also name the propery `images` and make it a list so that we won't need to change that if we add support for multiple images. In that case, in the future we could just add, perhaps, a new property `name` in the nested objects (next to `tag` and `location`) in order to let the deployment know which image is which.

lvarin replied:
For usability, when registering the project/repository there should be the possibility to set up a default Dockerfile inside the repository. But also allow here to change it. I feel most users will not change the location of the Dockerfile ever after registration.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+          type: string
+          description: Tag name for the created image
+          example: 'v1'
+        branch:
```
I think we should support `branch` or `tag`, and one of them should be required.

lvarin replied:
Yes, see below my train of thought :)

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+    BuildInfo:
+      type: object
+      properties:
+        commitSHA:
```
I think this should be optional when _registering_ a commit in `POST` and always point to the head commit by default. The specific commit hash should then be filled in by the broker so that it is always available when `GET`ting the information of a specific build (just like the start/finish time and the build status).

uniqueg replied:
Also, this is not relevant/applicable when dealing with a `tag` rather than a branch. So perhaps it may be a good idea to nest this object differently.

lvarin replied:
Thinking about this, I believe the most flexible approach is to change it to `ref`, a ref in Git can be a `commit`, a `branch` or a `tag`. If I put my "user of this broker" hat I will never use `commit`, only `branch` for feature/development branches, and `tag` for stable/prod branches. It is nice to allow `commit` for corner cases (or standard ones I cannot imagine).

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+            type: string
+            description: URL on which CD is listening for image updates.
+            example: https://ec2-54-203-145-132.compute-1.amazonaws.com/update
+    BuildInfo:
```
Here's a proposal (not tested!) for the entire `Build` schema (already incorporates several comments from above and below):

```yaml
Build:
  type: object
  description: # add description
  properties:
    status:
      $ref: '#/components/schemas/BuildStates'
    started_at:
      type: string
      format: date-time
      description: # add description
    finished_at:
      type: string
      format: date-time
      description: # add description
  allOf:
    - $ref: '#/components/schemas/BuildRegister'
  required:
    - status
    - images
    - head_commit
```

with

```yaml
BuildStates:
  type: string
  enum:
    - UNKNOWN
    - QUEUED
    - RUNNING
    - DELETING
    - SUCCEEDED
    - FAILED
```

and

```yaml
BuildRegister:
  type: object
  description: # add description
  properties:
    images:
      type: array
      items: $ref: '#/components/schemas/Image'
      description: # add description
    head_commit:
      $ref: '#/components/schemas/HeadCommit'
    required:
      - images
      - head_commit
```

and

```yaml
Image:
  type: object
  description: # add description
  properties:
    name:  # maybe better to avoid confusion with Git tags
      type: string
      description: # add description
    location:
      type: string
      description: # add description
      default: ./Dockerfile
    required:
      - name
```

and

```yaml
HeadCommit:
  oneOf:
    branch:
      type: object
      description: # add description
      properties:
        name:
          type: string
          description: # add description
        commit_sha:
          type: string
          description: # add description; describe that the head commit will be used if not supplied
        required:
          - name
    tag:
      type: string
      description: # add description
```

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+    Project:
+      type: object
+      properties:
+        type:
```
I think it would make sense to bundle different types of the same repo/project together. This should make it easier to register multiple types at the same time for the developer, and to explore the different available deployment types for administrators.

So that would lead to new endpoint routes of the form:

- `/projects/{id}`
- `/projects/{id}`/types/{type_id}`
- `/projects/{id}`/types/{type_id}/builds/{build_id}`

I would also propose that each type can be freely chosen by a developer (we can still define guidelines how this should ideally look like within ELIXIR, but also accept if some projects, for various reasons, do not stick to those) comes with a _mandatory_ description provided by the developer so that the administrator has a better idea which one to choose.

akash2237778 replied:
@uniqueg I'm not able to understand this completely. Do you mean different repositories of a similar type? Like we have _drs-filler_ and _trs-filler_ at ELIXIR?

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+          type: string
+          description: Development type
+          example: production
+        url:
```
Since we clearly require a _repository_ URL here, I would rename `projects` to `repositories` for clarity (see comment above).

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+          type: string
+          description: URL to the code repository
+          example: https://github.com/elixir-cloud-aai/trs-filer.git
+        dockerfileLocation:
```
This might change with each update, so should probably not be specified at the level of project registration. There may also be multiple Dockerfiles. See above for further comments.

lvarin replied:
This can be useful only if it is optional

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+          type: string
+          description: Location of container registry to push build image
+          example: gcr.io/my-project/busybox
+        containerRegistryToken:
```
Please have a look what the security implications are of posting a token in the request body in clear text.

lvarin replied:
AFAIK If the API uses https and never http, the request should be also encrypted. The main problem here is to store the token in clear text, but I do not think there is anything we can do other than ask the user for a dedicated token, i.e. Do not share this token with any other system.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+          type: string
+          description: Location of Dockerfile relative to repo root dir
+          example: docker/Dockerfile
+        containerRegistryUrl:
```
Consider to have this optional and set up a default container registry the broker will use if not supplied.

lvarin replied:
I shouldn't be complicated to have an internal registry, it is very easy to deploy. The only drawbacks I can think of are storage quotas and backups.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+            $ref: '#/components/schemas/Error'
+  
+  schemas:
+    Service:
```
Check the schemas carefully with regard to `required` properties. Also, please consider setting the `additionalProperties` flag defined in OpenAPI in some/all schemas to prohibit requests from adding additional properties. This may, however, mess with the use of `allOf` and `oneOf` properties.

uniqueg commented on broker.yaml:
```diff
@@ -0,0 +1,728 @@
+          type: string
+          description: Token to access container registry
+          example: XXXXXXXXXXXXXXX
+    Error:
```
Please have a look and try to make this compliant with https://datatracker.ietf.org/doc/html/draft-ietf-appsawg-http-problem-00 in case it is not. Also, ideally incorporate any recommendations (`SHOULD`) from that spec, as well as best practices (e.g., you could include the `type` and `instance` properties as described).

uniqueg reviewed (changes requested):
Hi Akash, despite a TON of comments, this is a very good first pass at the specs! :) I have been intentionally picky with this first review because I think clarity and unambiguity are extremely important for writing specifications, and consistency (which quite a few comments deal with) and precision of descriptions is very important to achieve that. If the specs are really solid, writing the app should be mostly a breeze, plus the service will be mostly auto-documented already.

Reviewing also has revelaled a couple of more conceptual issues that need further discussion. If things are complicated but not essential for a first MVP, please move them to issues for now. We can then deal with these later. It is, however, important that we get the routes right, otherwise later changes will be too extensive and we will lose time. To achieve this, I think the first priority is to discuss, in some more detail, the deployment types (who defines them? GA4GH? ELIXIR? Developers? Admins?). This involves checking the GA4GH Service Registry specs (do they define something like types, e.g., `production`, `development`) in some enum and the ELIXIR Cloud Service Registry additions to it. After that, it would be good to have a list of options, each with advantages and disadvantes listed.

lvarin commented:
I will check this better tomorrow. There are some things I did not have time to read and other that made think :)
I agree with 90% of the changes suggested

uniqueg commented:
Thanks @lvarin

lvarin commented:
Other than the comments I just added, I agree with all the changes and suggestions. What do you think @akash2237778

akash2237778 reviewed (commented):
> Other than the comments I just added, I agree with all the changes and suggestions. What do you think @akash2237778

Yes, that's good. I am working on comments, and probably by tonight, I'll be able to resolve them. Thanks for the review :)

akash2237778 committed:
Updated specs according to reviews

akash2237778 committed:
Updated specs - added additionalProperties and required tags.

akash2237778 committed:
Update specs

akash2237778 committed:
fix: connexion.exceptions.InvalidSpecification

akash2237778 committed:
Minor updates

uniqueg commented:
@akash2237778: I think everything in this PR is also part of PR #8, correct? If so, I think we can just close this.

uniqueg reviewed (commented):
This PR looks good but may be redundant because of #8. Please confirm, @akash2237778
